src/emitter.ts

```diff
@@ -0,0 +1,1126 @@
+        const conflictedMembers = extendConflictsBaseTypes[i.name] ? extendConflictsBaseTypes[i.name].memberNames : new Set();
+        emitProperties(prefix, emitScope, i, conflictedMembers);
+        const methodPrefix = prefix.startsWith("declare var") ? "declare function " : "";
+        emitMethods(methodPrefix, emitScope, i, conflictedMembers);
+    }
+
+    /// Emit all members of every interfaces at the root level.
+    /// Called only once on the global polluter object
+    function emitAllMembers(i: Browser.Interface) {
+        const prefix = "declare var ";
+        emitMembers(prefix, EmitScope.All, i);
+
+        for (const relatedIName of iNameToIDependList[i.name]) {
+            const i = allInterfacesMap[relatedIName];
+            if (i) {
+                emitAllMembers(i);
+            }
+        }
+    }
+
+    function emitEventHandlers(prefix: string, i: Browser.Interface) {
+        function getOptionsType(addOrRemove: string) {
+            return addOrRemove === "add" ? "AddEventListenerOptions" : "EventListenerOptions";
+        }
+
+        const fPrefix = prefix.startsWith("declare var") ? "declare function " : "";
+
+        function emitTypedEventHandler(prefix: string, addOrRemove: string, iParent: Browser.Interface) {
+            printer.printLine(`${prefix}${addOrRemove}EventListener<K extends keyof ${iParent.name}EventMap>(type: K, listener: (this: ${i.name}, ev: ${iParent.name}EventMap[K]) => any, options?: boolean | ${getOptionsType(addOrRemove)}): void;`);
+        }
+
+        function emitStringEventHandler(addOrRemove: string) {
+            printer.printLine(`${fPrefix}${addOrRemove}EventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | ${getOptionsType(addOrRemove)}): void;`);
+        }
+
+        function tryEmitTypedEventHandlerForInterface(addOrRemove: string) {
+            if (iNameToEhList[i.name] && iNameToEhList[i.name].length) {
+                emitTypedEventHandler(fPrefix, addOrRemove, i);
+                return true;
+            }
+            if (iNameToEhParents[i.name] && iNameToEhParents[i.name].length) {
+                iNameToEhParents[i.name]
+                    .sort(compareName)
+                    .forEach(i => emitTypedEventHandler(fPrefix, addOrRemove, i));
+                return true;
+            }
+            return false;
+        }
+
+        function emitEventHandler(addOrRemove: string) {
+            if (tryEmitTypedEventHandlerForInterface(addOrRemove)) {
+                // only emit the string event handler if we just emited a typed handler
+                emitStringEventHandler(addOrRemove);
+            }
+        }
+
+        emitEventHandler("add");
+        emitEventHandler("remove");
+    }
+
+    function emitConstructorSignature(i: Browser.Interface) {
+        const constructor = typeof i.constructor === "object" ? i.constructor : undefined;
```

Comment:
Was there some bad input this is avoiding?

Reply:
no this is cause all objects have a `constructor` property on them :D

Reply:
Oh yeah :man_facepalming: